### PR TITLE
Navigation/treemap navigation

### PR DESCRIPTION
**Pull request by Beta-Alf**
Adds a new navigation technique that improves navigation in treemaps and utilizes a combined projection.

### CODE REVIEW

File: source/gloperate-qt/include/gloperate-qt/viewer/Viewer.h [hunk 1]
```diff
@@ -172,7 +172,7 @@ protected slots:
     std::unique_ptr<PluginApi>                       m_pluginApi;
 
     std::unique_ptr<gloperate::Painter>              m_painter;
-    std::unique_ptr<DefaultMapping>                  m_mapping;
+    std::unique_ptr<TreeNavigationMapping>                  m_mapping;
```
Comment (p-otto) [inline source/gloperate-qt/include/gloperate-qt/viewer/Viewer.h]: Whitespace

File: source/gloperate-qt/source/viewer/TreeNavigation.cpp [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+    static const float DEFAULT_DISTANCE   = 2.0f;
+    static const float DEFAULT_DIST_MIN   = 0.1f;
+    static const float DEFAULT_DIST_MAX   = 4.0f;
+
+    static const float ROTATION_HOR_DOF   = 0.8f * glm::pi<float>();
+    static const float ROTATION_VER_DOF   = 0.8f * glm::pi<float>();
+
+    static const float ROTATION_KEY_SCALE = 1.0f;
+
+    //static const float NAV_CONSTRAINT_PAN_CIRCLE_R = 2.83;
```
Comment (p-otto) [inline source/gloperate-qt/source/viewer/TreeNavigation.cpp]: Code in comment

File: source/gloperate-qt/source/viewer/TreeNavigation.cpp [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+    return m_mode;
+}
+
+void TreeNavigation::reset()
+{
+    m_cameraCapability.reset();
+
+    m_mode = NoInteraction;
+
+//    enforceWholeMapVisible();
```
Comment (p-otto) [inline source/gloperate-qt/source/viewer/TreeNavigation.cpp]: Code in comment

File: source/gloperate-qt/source/viewer/TreeNavigation.cpp [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+        m_eye = m_cameraCapability.eye();
+        m_center = m_cameraCapability.center();
+        
+        auto up = m_cameraCapability.up();
+        auto viewDir = glm::normalize(m_eye - m_center);
+        auto va = acosf(glm::dot(viewDir, up));
+        
+        auto tween_va = (va - CONSTRAINT_ROT_MAX_V_UP) / PROJECTION_TWEENING_THRESH;
+        tween_va = glm::clamp(tween_va, 0.0f, 1.0f);
+        //tween_va = tween_va < 0.5f ? 0.0f : 1.0f;
```
Comment (p-otto) [inline source/gloperate-qt/source/viewer/TreeNavigation.cpp]: More outcommented code

File: source/gloperate-qt/source/viewer/TreeNavigation.cpp [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+{
+    const glm::vec3 ln = m_cameraCapability.eye();
+    const glm::vec3 lf = m_cameraCapability.center();
+
+    bool intersects = true;
+    glm::vec3 i = navigationmath::rayPlaneIntersection(intersects, ln, lf);
+    if (!intersects)
+        return;
+
+    //enforceScaleConstraints(scale, i);
```
Comment (p-otto) [inline source/gloperate-qt/source/viewer/TreeNavigation.cpp]: +

File: source/gloperate-qt/source/viewer/TreeNavigation.cpp [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+{
+    // retrieve the angle between camera-center to up and test how much closer
+    // to up/down it can be rotated and clamp if required.
+    auto up = m_cameraCapability.up();
+    auto viewDir = glm::normalize(m_eye - m_center);
+    auto horizontalDir = glm::normalize(viewDir - (up * glm::dot(viewDir, up)));
+
+    auto ha = acosf(glm::dot(m_cardinalDirection, horizontalDir));
+    auto hDist = CONSTRAINT_ROT_MAX_H-ha;
+    /*if(std::signbit(hDist))
```
Comment (p-otto) [inline source/gloperate-qt/source/viewer/TreeNavigation.cpp]: +

File: source/gloperate-qt/source/viewer/TreeNavigationMapping.cpp [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+    m_typedRenderTargetCapability = nullptr;
+    m_coordProvider = nullptr;
+    m_navigation = nullptr;
+
+    if (!m_painter)
+    {
+        return;
+    }
+
+    if (m_painter->supports<AbstractCameraCapability>() &&
```
Comment (p-otto) [inline source/gloperate-qt/source/viewer/TreeNavigationMapping.cpp]: Possible to move some logic to bool variables?

File: source/gloperate-qt/source/viewer/TreeNavigationMapping.cpp [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+    default:
+        break;
+    }
+}
+
+void TreeNavigationMapping::mapKeyboardEvent(KeyboardEvent * event)
+{
+    if (event && event->type() == KeyboardEvent::Type::Press)
+    {
+        switch (event->key())
```
Comment (p-otto) [inline source/gloperate-qt/source/viewer/TreeNavigationMapping.cpp]: Change to map from key to action?

File: source/gloperate/source/painter/CombinedProjectionCapability.cpp [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+namespace gloperate{
+
+CombinedProjectionCapability::CombinedProjectionCapability(gloperate::AbstractViewportCapability *viewportCapability) 
+: gloperate::AbstractProjectionCapability(viewportCapability)
+, m_orthoCapability(viewportCapability)
+, m_perspectiveCapability(viewportCapability)
+, m_mix(1.0f)
+{
+
```
Comment (p-otto) [inline source/gloperate/source/painter/CombinedProjectionCapability.cpp]: Whitespace

File: source/gloperate-qt/source/viewer/TreeNavigationMapping.cpp [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+            break;
+        }
+    }
+}
+
+void TreeNavigationMapping::mapWheelEvent(WheelEvent * wheelEvent)
+{
+    auto scale = wheelEvent->angleDelta().y;
+    scale /= WheelEvent::defaultMouseAngleDelta();
+    scale *= 0.1f; // smoother (slower) scaling
```
Comment (p-otto) [inline source/gloperate-qt/source/viewer/TreeNavigationMapping.cpp]: Move magic number to variable

File: source/gloperate-qt/source/viewer/TreeNavigation.cpp [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+{
+    if (RotateInteraction != m_mode)
+        return;
+
+    const glm::vec2 delta(m_m0 - mouse);
+    // setup the degree of freedom for horizontal rotation within a single action
+    const float wDeltaX = delta.x / static_cast<float>(m_viewportCapability.width());
+    // setup the degree of freedom for vertical rotation within a single action
+    const float wDeltaY = delta.y / static_cast<float>(m_viewportCapability.height());
+
```
Comment (p-otto) [inline source/gloperate-qt/source/viewer/TreeNavigation.cpp]: Whitespace

File: source/gloperate-qt/include/gloperate-qt/viewer/TreeNavigation.h [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+    *    scale > 0: zoom in
+    */
+    void scaleAtMouse(const glm::ivec2 & mouse,	float scale);
+    void resetScaleAtMouse(const glm::ivec2 & mouse);
+
+//	void enforceWholeMapVisible(const float offset = 0.08);
+
+    // constraints
+
+//	void enforceTranslationConstraints(glm::vec3 & p) const;
```
Comment (p-otto) [inline source/gloperate-qt/include/gloperate-qt/viewer/TreeNavigation.h]: Code in comment

File: source/gloperate-qt/source/viewer/TreeNavigation.cpp (hunk 11 above)
Comment (Beta-Alf) [inline source/gloperate-qt/source/viewer/TreeNavigation.cpp]: Concerns following lines, was fixed

**Pull request merged**